wandb/sdk/internal/internal_api.py

```diff
@@ -1552,18 +1552,10 @@ def download_write_file(self, metadata, out_dir=None):
 
     def upload_file_azure(self, url, file, extra_headers):
         from azure.core.exceptions import HttpResponseError
-        from azure.core.pipeline.policies import HTTPPolicy
-
-        # Disable retries
-        class NoRetry(HTTPPolicy):
-            def send(self, request):
-                return self.next.send(request)
-
-            def increment(self, *args, **kwargs):
-                return 0
 
+        # Configure the client with a single retry
         client = self._azure_blob_module.BlobClient.from_blob_url(
-            url, retry_policy=NoRetry()
+            url, retry_policy=self._azure_blob_module.LinearRetry(retry_total=1)
```

Comment:
What happens if this fails after a retry? I mean is it informative enough so that we don't need to catch the exception.

Reply:
I was trying to avoid having two layers of retry logic.  When the azure sdk retry fails it will raise an `HttpResponseError` which we catch and then convert into a regular `requests` exception that should be caught and retried by our existing logic.

Reply:
I initially set `retry_total=0` and saw a read timeout error that didn't seem to get caught by our retry logic...  I'm guessing that means the azure sdk won't always throw an `HttpResponseError`, I'm going to do some more digging.  It's really hard to test stuff like this.  The unit tests assert the exception logic get's triggered but it's not trying all of the possible failure modes.

Reply:
@dmitryduev I figured out what was going on.  Just pushed a more robust fix.  The Azure SDK will only throw an `HTTPResponseError` if it got a response.  I now catch all `AzureError` instances and raise a `requests.exceptions.ConnectionError` if I don't have a response.  I was able to verify this works by toggling my wifi on and off during a run.

Reply:
I first laughed about your creative way of testing this @vanpelt, but having thought a bit more about it, I think we should totally add a similar capability to `yea`! Like, `subprocess.run(["sudo", "ifconfig", "eth0", "down"])` or something during a test (and then back up). @raubitsj?

Reply:
the testing plan was to incorporate mock_requests:InjectRequests() into yea.
So you can specify requests match rules in yaml format in your yea specification like:
```
inject:
   requests:
      - match:
           path_suffix: /blah
           count: 3
        http_status: 404
```

Reply:
I tested this previously here: https://github.com/wandb/client/blob/b9b9721837d7322c2b900c81a8e686858c9caaa6/tests/test_file_upload.py#L36

The test didn't catch the regression because our MockRequest object was too forgiving of what keyword arguments it allowed to session.request

It would be nice to test the connection error (read timeout, or network fault).